test/yaml_edit/test_case.dart

```diff
@@ -0,0 +1,318 @@
+        var inputName = inputUri.toFilePath(windows: false).split('/').last;
+        var inputNameWithoutExt = inputName.substring(0, inputName.length - 5);
+        var goldenUri = goldDirUri.resolve('./$inputNameWithoutExt.golden');
+
+        return TestCase(inputUri, goldenUri);
+      }).toList();
+    }
+
+    return TestCases(testCaseList);
+  }
+
+  /// Tests all the [TestCase]s if the golden files exist, create the golden
+  /// files otherwise.
+  void test() {
+    var tested = 0;
+    var created = 0;
+
+    for (var testCase in testCases) {
+      testCase.testOrCreate();
+      if (testCase.state == TestCaseStates.testedGoldenFile) {
+        tested++;
+      } else if (testCase.state == TestCaseStates.createdGoldenFile) {
+        created++;
+      }
+    }
+
+    print(
+        'Successfully tested $tested inputs against golden files, created $created golden files');
+  }
+
+  TestCases(this.testCases);
+
+  int get length => testCases.length;
+}
+
+/// Enum representing the different states of [TestCase]s.
+enum TestCaseStates { initialized, createdGoldenFile, testedGoldenFile }
+
+/// Interface for a golden test case. Handles the logic for test conduct/golden
+/// test update accordingly.
+class TestCase {
+  final Uri inputUri;
+  final Uri goldenUri;
+  final List<String> states = [];
+
+  String info;
+  YamlEditor yamlBuilder;
+  List<YamlModification> modifications;
+
+  String inputLineEndings = '\n';
+
+  TestCaseStates state = TestCaseStates.initialized;
+
+  TestCase(this.inputUri, this.goldenUri) {
+    var inputFile = File.fromUri(inputUri);
+    if (!inputFile.existsSync()) {
+      throw Exception('Input File does not exist!');
+    }
+
+    _initialize(inputFile);
+  }
+
+  /// Initializes the [TestCase] by reading the corresponding [inputFile] and
+  /// parsing the different portions, and then running the input yaml against
+  /// the specified modifications.
+  ///
+  /// Precondition: [inputFile] must exist, and inputs must be well-formatted.
+  void _initialize(File inputFile) {
+    var input = inputFile.readAsStringSync();
+
+    inputLineEndings = detectWindowsLineEndings(input) ? '\r\n' : '\n';
+    var inputElements = input.split('---$inputLineEndings');
+
+    if (inputElements.length != 3) {
+      throw AssertionError('File ${inputFile.path} is not properly formatted.');
+    }
+
+    info = inputElements[0];
+    yamlBuilder = YamlEditor(inputElements[1]);
+    var rawModifications = getValueFromYamlNode(loadYaml(inputElements[2]));
+    modifications = parseModifications(rawModifications);
+
+    /// Adds the initial state as well, so we can check that the simplest
+    /// parse -> immediately dump does not affect the string.
+    states.add(yamlBuilder.toString());
+
+    _performModifications();
+  }
+
+  void _performModifications() {
+    for (var mod in modifications) {
+      _performModification(mod);
+      states.add(yamlBuilder.toString());
```

Comment:
Just an idea:

We could consider catching `PathError` and `AliasError` and then DUMP these into `states` as `# throws PathError`....

That way we can change path-error, and alias errors into the golden tests..

Note. clearly we should check that this can only happen in the modifucation..

Reply:
hm I'm not sure if a golden test is the best place to check for path errors and alias errors - the way I see it, golden tests allow us best to easily compare the output space for space, character for character, for complicated or long series of changes, whereas unit testing in the test files might be more suited for the errors?

Reply:
Good point!